Tabulation refactor

mikefranze opened the pull request:
- Big rework of tabulation code to match starpy a bit more and implement new tie breaker logic. Old code assumed input format was unknown, so a lot of logic dealing with that was sprinkled throughout. This update assumes the input structure since we're defining it. If we end up reading files where we have to parse the data types we can do that in a separate function and then pass it into the tabulation code. 
- True ties are currently resolved randomly and reported in the results logs.
- Added test cases for new STAR tabulator. Test cases for true ties are bit janky at the moment as the only indication is in the logs.
- Added plurality and approval tabulator and hooked them up in the backend, was pretty straightforward. Don't have any support for front end however. 

The tabulation updates should eventually get merged into the star-core repo, and then we try importing them with npm again, but last time I tried I had a few issues.

scottPlusPlus commented on backend/src/Tabulators/StarResults.js:
```diff
@@ -1,759 +1,253 @@
-        noPrefVotes: noPrefVotes
-      };
+        else {
+          const randomWinner = scoreTieWinners[getRandomInt(scoreTieWinners.length)]
```
Not a code issue:  this seems like a huge problem.  Theoretically it is cheap to "run" the final round with either of the random winners.  Then we can present both outcomes  (if either runner up would lose, then simple.  but if either of them would potentially win the election???  would be good to surface that information clearly to the organizers)

From a code perspective:  if we are going to be using random for something important, we would want organizers to be able to pass in their own "random number" or whatever so they can have more trust here.  (or again, we could spit out the possible options as the result and let the organizers flip a coin or whatever)

mikefranze replied:
Seems like a pretty rare scenario, but we could spend a lot of time going down rabbit holes of various scenarios and Condorcet cycles. To keep the algorithm simple though we decided that ties in the score round should be decided just by who is tied and not considering who has already qualified for the runoff. We do want the tiebreaking protocol to be defined ahead of time but I suppose we could have a setting for if we just report the tie or resolve randomly (or use the 5 star tiebreaker that I completely forgot to add..). If we do resolve randomly it would be a good idea to have the seed be provided somehow. Maybe we can use the election ID as a seed or something. 

Also... I missed another part where if the number of tied candidates is equal to or less than the number of open seats, then they are all elected.

scottPlusPlus commented on backend/src/Tabulators/StarResults.js:
```diff
@@ -1,759 +1,253 @@
-        support: new Array(6).fill(0)
-      };
-      candidates.push(candidate);
+const ParseData = require("./ParseData");
```
This is probably the most important file of the project :)

Would lean to convert the whole thing to Typescript and add more comments. Also random q: can we have it set up so that different precincts can crunch their data separately, and then combine to see the final result?

BUT: totally legit to add this to our TODO list

mikefranze replied:
I didn't initially convert to Typescript because I wanted to eventually push this back to the STAR Core repo and have it be its own javascript npm package. But after learning a bit more it should be easy to compile it back to javascript and offer both, right? 

On the precinct question, yes we can! The python implementation currently lets you optionally pass in the summaryData instead of the ballots. So we could run the election for each precinct, sum up the data, then pass that into the tabulation function to get the total results. That way you'd get details for each precinct and the final results. It wouldn't be too much work, but lets wait until we add more precinct support in other areas of the project first.

mikefranze commented on backend/src/Tabulators/Star.test.js:
```diff
@@ -0,0 +1,144 @@
+        ]
+        const results = StarResults(candidates, votes, 1)
+        const logs = results.round_results[0].logs
+        expect(logs[logs.length - 1].slice(0, 8)).toBe('True tie');//Need better tie reporting?
```
In the instances of true ties we pick a random winner and report that there was a tie in the results log. So we do get a winner, but to assert that there was a tie I had to parse the log. I'm thinking there's a better way to do this though.

mikefranze commented on backend/src/Tabulators/Star.test.js:
```diff
@@ -0,0 +1,144 @@
+        ]
+        const results = StarResults(candidates, votes, 1)
+        const logs = results.round_results[0].logs
+        expect(logs[logs.length - 1].slice(0, 8)).toBe('True tie');//Need better tie reporting?
```
Winner would be a random pick of the tied candidates. This just tests that a tie occurred.

mikefranze commented on backend/src/Tabulators/Star.test.js:
```diff
@@ -0,0 +1,144 @@
+    test("Condorcet Winner", () => {
+        const candidates = ['Allison', 'Bill', 'Carmen', 'Doug']
+        const votes = [
+            [5, 2, 1, 4],
```
We do, we have counts of valid votes, invalid votes, undervotes, and bullet votes. I can add a test for that.

mikefranze commented on backend/src/Tabulators/Util.js:
```diff
@@ -0,0 +1,35 @@
+
+// Format a Timestamp value into a compact string for display;
```
This is one of the old functions that I didn't want to throw away. Would only be used if we were parsing a csv with an unknown header or something. But I'll test it out and add some comments.

mikefranze commented:
Remaining TODOs:

- [ ] Convert to typescript
- [ ] Add 5 star tiebreaker
- [ ] Add logic to elect tied candidates if seats are available
- [ ] Add tiebreaker option to settings
- [ ] Tests for valid/invalid/undervote/bullet vote counts
- [ ] More comments

mikefranze commented:
Addressed most issues I think. Added settings to enable five-star tiebreaker and random tiebreaker which backend automatically sets to true. Will need to add it to election settings in the future. 
Also added first cut at new Ranked Robin method while I was working tabulators and added it to sandbox tool. Anything beyond basic tie breaker isn't added yet and just picks random winner to break ties.
